firedrake/ensemble/ensemble_function.py

```diff
@@ -0,0 +1,292 @@
+from firedrake.petsc import PETSc
+from firedrake.ensemble.ensemble_functionspace import (
+    EnsembleFunctionSpace, EnsembleDualSpace)
+from firedrake.adjoint_utils import EnsembleFunctionMixin
+from firedrake.function import Function
+from firedrake.norms import norm
+from pyop2 import MixedDat
+
+from functools import cached_property
+from contextlib import contextmanager
+
+__all__ = ("EnsembleFunction", "EnsembleCofunction")
+
+
+class EnsembleFunctionBase(EnsembleFunctionMixin):
+    """
+    A mixed (co)function defined on a :class:`firedrake.Ensemble`.
+    The subcomponents are distributed over the ensemble members, and
+    are specified locally in a :class:`firedrake.EnsembleFunctionSpace`.
+
+    Parameters
+    ----------
+
+    function_space : :class:`firedrake.EnsembleFunctionSpace`.
+        The function space of the (co)function.
+
+    Notes
+    -----
+    Passing a :class:`firedrake.EnsembleDualSpace` to :class:`firedrake.EnsembleFunction`
+    will return an instance of :class:`firedrake.EnsembleCofunction`.
+
+    This class does not carry UFL symbolic information, unlike a
+    :class:`firedrake.Function`. UFL expressions can only be defined
+    locally on each ensemble member using a `firedrake.Function`
+    from `EnsembleFunction.subfunctions`.
+
+    See also:
+    - Primal ensemble objects: :class:`firedrake.EnsembleFunctionSpace` and :class:`firedrake.EnsembleFunction`.
+    - Dual ensemble objects: :class:`firedrake.EnsembleDualSpace` and :class:`firedrake.EnsembleCofunction`.
+    """
+
+    @PETSc.Log.EventDecorator()
+    @EnsembleFunctionMixin._ad_annotate_init
+    def __init__(self, function_space):
+        self._fs = function_space
+
+        self._fbuf = Function(function_space._full_local_space)
```

Comment:
This isn't a very clear name

Reply:
It's an internal detail (`_`), its a `Function` (`f`), but really it's just a storage buffer (`buf`).
Something like `_full_local_function` might be better to be consistent with `function_space._full_local_space`.